### PR TITLE
[FSSDK-9605] add sendBeacon event dispatcher support

### PR DESCRIPTION
**Pull request by raju-opti**
## Summary
-  add support for event dispatcher using sendBeacon which will be used in browser on optimizely client close.

## Test plan
- all existing tests should pass
- added new tests for the new dispatcher and event processor
- manually tested that event are being sent using sendBeacon on close

## Issues
- FSSDK-9605

### CODE REVIEW

File: lib/index.browser.ts [hunk 1]
```diff
@@ -111,6 +115,7 @@ const createInstance = function(config: Config): Client | null {
 
     const eventProcessorConfig = {
       dispatcher: eventDispatcher,
+      closingDispatcher,
```
Comment (jaeopt) [inline lib/index.browser.ts:118]: Wondering if we see any possible side-effect and need to make this configurable.

File: lib/modules/event_processor/eventProcessor.ts [hunk 2]
```diff
@@ -57,13 +57,20 @@ export function validateAndGetBatchSize(batchSize: number): number {
...
 }
 
-export function getQueue(batchSize: number, flushInterval: number, sink: EventQueueSink<ProcessableEvent>, batchComparator: (eventA: ProcessableEvent, eventB: ProcessableEvent) => boolean ): EventQueue<ProcessableEvent> {
+export function getQueue(
+  batchSize: number, 
+  flushInterval: number, 
+  batchComparator: (eventA: ProcessableEvent, eventB: ProcessableEvent) => boolean,
+  sink: EventQueueSink<ProcessableEvent>,
+  closingSink?: EventQueueSink<ProcessableEvent>
+): EventQueue<ProcessableEvent> {
```
Comment (jaeopt) [inline lib/modules/event_processor/eventProcessor.ts:66]: Don't we need the same beacon support for ODPEventDispatcher as well for consistency?

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks good! A couple of clarifications.

File: lib/modules/event_processor/eventProcessor.ts (hunk 2 above)
Comment (raju-opti) [reply lib/modules/event_processor/eventProcessor.ts:66]: Currently, ODPEventManager for browser does not queue any events, it sends them immediately, so we don't have a similar use case.

File: lib/index.browser.ts (hunk 1 above)
Comment (raju-opti) [reply lib/index.browser.ts:118]: Yeah, probably should make it configurable as well.

File: lib/modules/event_processor/eventProcessor.ts (hunk 2 above)
Comment (jaeopt) [reply lib/modules/event_processor/eventProcessor.ts:66]: what is expected if we have back-2-back ODP events? 2nd one is discarded or blocking?

Comment (raju-opti) [reply lib/modules/event_processor/eventProcessor.ts:66]: @jaeopt in that case both event will be sent immediately asynchronously, without blocking or waiting.

**Comment by alexparish** [comment]
This is really great. 

In order to synchronise activations to a data platform I have created a custom event dispatcher that I pass to `createInstance`. Ultimately this custom event dispatcher calls `eventDispatcher.dispatchEvent(eventObject, callback)` with `eventDispatcher` imported from the SDK package.

Will this pull request support my use case or will I have to integrate `sendBeaconDispatcher` into my custom event dispatcher?

**Comment by raju-opti** [comment]
> This is great, thanks for implementing the Beacon API.
> 
> In order to synchronise activations to a data platform I am using a custom event dispatcher that I pass to `createInstance`. Ultimately this custom event dispatcher calls `eventDispatcher.dispatchEvent(eventObject, callback)` with `eventDispatcher` imported from the SDK package.
> 
> Will people like myself using custom event dispatchers need to manually integrate`sendBeaconDispatcher`?

@alexparish yes, you will need to wrap the `sendBeaconEventDispatcher` similarly and pass it as `closingEventDispatcher` option to `createInstance`.

**Comment by mikechu-optimizely** [comment]
@raju-opti and I are finishing our manual e2e acceptance testing (bug bashing) after the code review....

**Review by mikechu-optimizely — COMMENTED** [review]
Some questions

File: lib/modules/event_processor/eventQueue.ts [hunk 3]
```diff
@@ -14,149 +14,149 @@
...
     batchComparator,
   }: {
-    flushInterval: number
-    maxQueueSize: number
-    sink: EventQueueSink<K>
-    batchComparator: (eventA: K, eventB: K) => boolean
+    flushInterval: number;
+    maxQueueSize: number;
+    sink: EventQueueSink<K>;
+    closingSink?: EventQueueSink<K>;
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/modules/event_processor/eventQueue.ts:103]: This is just a field in an object, so order should not matter.

File: lib/modules/event_processor/v1/v1EventProcessor.ts [hunk 4]
```diff
@@ -36,31 +36,41 @@ const logger = getLogger('LogTierV1EventProcessor')
...
 
   constructor({
     dispatcher,
+    closingDispatcher,
     flushInterval = DEFAULT_FLUSH_INTERVAL,
     batchSize = DEFAULT_BATCH_SIZE,
     notificationCenter,
   }: {
     dispatcher: EventDispatcher
+    closingDispatcher?: EventDispatcher
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/modules/event_processor/v1/v1EventProcessor.ts:52]: Order of optional arguments to a function matters in TS, they must come after all mandatory params. However, here closingDispatcher is a field of an object, not a parameter of the constructor itself. So order should not matter. And as it is an optional field, the change to API is non-breaking.

File: lib/plugins/event_dispatcher/send_beacon_dispatcher.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * Users of the SDK can provide their own implementation
+ * @param  {Event}    eventObj
+ * @param  {Function} callback
+ */
+export const dispatchEvent = function(
+  eventObj: Event,
+  callback: (response: { statusCode: number; }) => void
+): void {
+  const params = eventObj.params;
+  const url: string = eventObj.url;
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/plugins/event_dispatcher/send_beacon_dispatcher.ts]: updated

**Review by mikechu-optimizely — APPROVED** [review]
LGTM. 👍   Thanks.

**Comment by mikechu-optimizely** [comment]
@jaeopt, Are you good with this PR?

**Review by jaeopt — CHANGES_REQUESTED** [review]
Changes look good to me! 
With this change, sendBeacon is not default any more, so clients should instantiate the custom handler to use it. Other option is to use it as the default and they can turn it off if they do not like it. WDYT?

**Comment by raju-opti** [comment]
> Changes look good to me! With this change, sendBeacon is not default any more, so clients should instantiate the custom handler to use it. Other option is to use it as the default and they can turn it off if they do not like it. WDYT?

@jaeopt As this PR introduces option to use a separate dispatcher when closing the instance, this would break users using custom event dispatcher if used by default when closing, that's why we have to keep it optional.

**Comment by alexparish** [comment]
Is there a reason why events cannot always be sent using `sendBeacon` in the browser? It would simplify the change, for example `closingEventDispatcher` would not need to be specified when using `createInstance`,.

**Comment by raju-opti** [comment]
> Is there a reason why events cannot always be sent using `sendBeacon` in the browser? It would simplify the change, for example `closingEventDispatcher` would not need to be specified when calling `createInstance`.

Yes, sendBeacon does not provide any way of checking the HTTP response status. By default, we will want to check the response status and retry in case of 5xx status.

**Comment by mikechu-optimizely** [comment]
Last commit, still LGTM.

**Review by jaeopt — APPROVED** [review]
LGTM

**Comment by mikechu-optimizely** [comment]
Woot 🎉

**Pull request merged**